### PR TITLE
feat(integrations): add pagerduty integration (#2560)

### LINKED ISSUE #2560: feat: add PagerDuty integration
## Context

PagerDuty is the most widely used on-call and incident management platform in the SRE ecosystem. OpenSRE currently covers OpsGenie but has no PagerDuty support, which is a significant gap for teams that route all their alerts through PagerDuty.

## What to build

- Fetch active incidents and their timeline
- Fetch on-call schedules and current on-call responders
- Fetch services, escalation policies, and alert routing rules
- Correlate PagerDuty alerts with infrastructure events during RCA

## Acceptance Criteria

Follow the [New Integration Checklist](https://github.com/Tracer-Cloud/opensre/blob/main/AGENTS.md#6-new-integration-checklist):

- [ ] Integration config and normalization under `app/integrations/pagerduty.py`
- [ ] Service client under `app/services/pagerduty/client.py` with proper timeout handling
- [ ] At least one tool under `app/tools/PagerDutyTool/` with typed input schema
- [ ] Unit/mock tests under `tests/integrations/test_pagerduty.py`
- [ ] Docs page at `docs/integrations/pagerduty.mdx` registered in `docs/docs.json`
- [ ] Synthetic or e2e test covering the core tool behavior
- [ ] `make verify-integrations` passes
- [ ] `make lint` and `make typecheck` pass
- [ ] **Video demo (required):** screen recording showing an investigation that surfaces PagerDuty incident data as part of the RCA output — must be attached to the PR before review

### PR DESCRIPTION
**Pull request by kasulani**
Fixes #2560 

#### Describe the changes you have made in this PR
Adds full PagerDuty integration to OpenSRE, enabling the investigation agent to pull incident data, on-call schedules, and service topology from PagerDuty during root cause analysis.

##### Setup integration (happy path)

https://github.com/user-attachments/assets/eabef6c3-6864-4e8b-a9a0-48ce2a3f76e8

##### Setup integration (unhappy path)
https://github.com/user-attachments/assets/ea487b85-d354-4867-9e9b-fce5759b7e32

##### Using the chat /investigate command
The recording for this is large so I decide to paste the terminal output

```
╭─  OpenSRE · v0.1  ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                                                                                     │
│  █▀█                                              │ Integrations                                                                                                    │
│  █▄█                                              │ run /onboard to connect tools                                                                                   │
│                                                   │ ───                                                                                                             │
│                                                   │ Alert listener                                                                                                  │
│  Welcome back eka!                                │ ○  not configured                                                                                               │
│                                                   │                                                                                                                 │
│                                                   │                                                                                                                 │
│  gpt-5.4-mini  ·  openai  ·                       │                                                                                                                 │
│  ~/Desktop/personal/opensre                       │                                                                                                                 │
│                                                                                                                                                                     │
╰─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯

❯ /investigate .idea/pagerduty-incident.csv

0:00  ◐  READ    ·  Loading integrations

0:00  ✓  READ    ·  Loading integrations  pagerduty  0ms

0:00  ◐  READ    ·  Reading alert

  ┃  Alert     High CPU on service-test
  ┃  Pipeline  service-test
  ┃  Severity  high


0:02  ✓  READ    ·  Reading alert  2.9s

0:04  ◐  INVEST  ·  Investigation

0:20  ✓  INVEST  ·  Investigation  16.8s  ↳ Hermes: log poll | Pagerduty: incidents, services, incident detail

0:20  ◐  DIAG    ·  Correlate Upstream

0:20  ✓  DIAG    ·  Correlate Upstream  0ms

  The incident was caused by sustained CPU saturation on `service-test`, with PagerDuty explicitly reporting CPU remained above 90% long enough to trigger and escalate
a P1 incident

 Findings  ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

    · `service-test` currently has an active PagerDuty incident.
    · The incident title states: `High CPU on service-test — sustained above 90%`.
    · The incident was created at `2026-06-04T07:57:57Z`.
    · The incident was still active at `2026-06-04T08:28:00Z`.
    · The incident severity was elevated to `P1`.
    · The trigger source was `web_trigger`.
    · No Hermes log records were returned from the checked log file.
  *Non-Validated Claims (Inferred):*
    · The exact process, endpoint, or workload causing the CPU spike is unknown.
    · The spike may have been caused by a deployment, traffic increase, infinite loop, lock contention, or a large job, but this is unconfirmed.
    · Whether the issue was resolved after the last PagerDuty update is unknown.
    · Whether the service has a history of recurring hotspots is unknown.

 Recommended Actions  ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

    · Inspect CPU utilization and per-process/thread profiles for `service-test` around `2026-06-04T07:57Z`.
    · Check recent deploys, config changes, and traffic spikes in the same time window.
    · If a single worker or shard is overloaded, rebalance or shard the workload more evenly.
    · Temporarily reduce concurrency or request rate, or scale out capacity if available.
    · Add monitoring for per-process CPU, queue depth, and hotspot detection.
    · If a particular input or job is pathological, add safeguards to skip, split, or throttle it.

 Cited Evidence  ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

    — Queries: get hermes logs, pagerduty incidents, pagerduty services, pagerduty incident detail, get sre guidance
```
---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [ ] No, I wrote all the code myself
- [x] Yes, I used AI assistance (continue below)

**If you used AI assistance:**
- [x] I have reviewed every single line of the AI-generated code
- [x] I can explain the purpose and logic of each function/component I added
- [x] I have tested edge cases and understand how the code handles them
- [x] I have modified the AI output to follow this project's coding standards and conventions

**Explain your implementation approach:**

**Problem:** OpenSRE had no way to pull PagerDuty context during investigations — incidents, responder status, and service topology were invisible to the agent.

**Approach:** Followed the established OpsGenie integration pattern (closest existing parallel) — dedicated service client, BaseTool subclasses with `is_available`/`extract_params`/`run`, and the standard catalog/registry/verification wiring.

**Why this shape:** The tool registry auto-discovers BaseTool subclasses, so no pipeline changes were needed. The client is a thin httpx wrapper that the tools instantiate per-call via `make_pagerduty_client`, keeping tool execution stateless and thread-safe.

**Key components:**
- `PagerDutyClient` — handles auth headers, timeouts, and error capture for all API interactions
- 4 tools — each focused on one query type the investigation planner can select independently
- Catalog/registry wiring — makes PagerDuty discoverable via env vars or the local store without touching the pipeline

---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issue
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [x] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

---

Note: Please check **Allow edits from maintainers** if you would like us to assist in the PR.

### CODE REVIEW

File: tests/services/pagerduty/test_client.py [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+        assert c._client is not None
+    assert c._client is None
+
+
+def test_context_manager_closes_on_exception() -> None:
+    c = _client()
+    _ = c._get_client()
+    with pytest.raises(ValueError), c:
+        raise ValueError("test error")
+    assert c._client is None
```
[the remark this answers is not in the document]
Comment (kasulani) [reply tests/services/pagerduty/test_client.py:442]: This is a false positive. CodeQL does not understand `with pytest.raises(ValueError), c:` construct. This is equivalent to:
```
with pytest.raises(ValueError):
    with c:
        raise ValueError("test error")
```
The linter rejects the above construct.

**Comment by kasulani** [comment]
@greptile review

**Comment by kasulani** [comment]
@greptile review

**Comment by kasulani** [comment]
@greptile review

**Comment by kasulani** [comment]
@muddlebee this PR is ready for review

**Comment by kasulani** [comment]
@greptile review

**Comment by Davidson3556** [comment]
@kasulani  Looks good and mirrors the OpsGenie integration closely. One thing to fix first:

In app/integrations/registry.py the PagerDuty spec picks order numbers that are already used on main — setup_order=24 is taken by jenkins and dagster, and verify_order=24 is taken by bitbucket. When two services share a number, their order in the wizard and in `integrations verify` is undefined. Can you bump these to free values? 25 and 37 are both open.

Minor: there's a blank-line deletion in app/integrations/cli.py that isn't related to PagerDuty: worth reverting to keep the diff tight.

**Comment by kasulani** [comment]
> @kasulani Looks good and mirrors the OpsGenie integration closely. One thing to fix first:
> 
> In app/integrations/registry.py the PagerDuty spec picks order numbers that are already used on main — setup_order=24 is taken by jenkins and dagster, and verify_order=24 is taken by bitbucket. When two services share a number, their order in the wizard and in `integrations verify` is undefined. Can you bump these to free values? 25 and 37 are both open.
> 
> Minor: there's a blank-line deletion in app/integrations/cli.py that isn't related to PagerDuty: worth reverting to keep the diff tight.

Hey @Davidson3556, I have fixed the issues you pointed out. Thanks for the feedback

**Comment by cerencamkiran** [comment]
@greptile review

**Comment by muddlebee** [comment]
@kasulani pushed some nits, otherwise good to go
